File: common/src/main/java/com/ssblur/scriptor/particle/ScriptorParticles.java
package com.ssblur.scriptor.particle;

import com.mojang.serialization.MapCodec;
import com.ssblur.scriptor.ScriptorMod;
import dev.architectury.platform.Platform;
import dev.architectury.registry.client.particle.ParticleProviderRegistry;
import dev.architectury.registry.registries.DeferredRegister;
import dev.architectury.registry.registries.RegistrySupplier;
import net.fabricmc.api.EnvType;
import net.minecraft.core.particles.ParticleType;
import net.minecraft.core.registries.Registries;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;

public class ScriptorParticles {
  public static final DeferredRegister<ParticleType<?>> PARTICLE_TYPES = DeferredRegister.create(ScriptorMod.MOD_ID, Registries.PARTICLE_TYPE);

  public static final RegistrySupplier<ParticleType<MagicParticleData>> MAGIC = PARTICLE_TYPES.register(
    "magic",
    () -> new ParticleType<>(true) {
      @Override
      public MapCodec<MagicParticleData> codec() {
        return MagicParticleData.CODEC;
      }

      @Override
      public StreamCodec<? super RegistryFriendlyByteBuf, MagicParticleData> streamCodec() {
        return MagicParticleData.STREAM_CODEC;
      }
    }
  );

  public static void register() {
    if(Platform.getEnv() == EnvType.CLIENT) {
      ParticleProviderRegistry.register(MAGIC, MagicParticle::new);
    }

    PARTICLE_TYPES.register();
  }
}

File: common/src/main/kotlin/com/ssblur/scriptor/particle/ScriptorParticles.kt
package com.ssblur.scriptor.particle

import com.mojang.serialization.MapCodec
import com.ssblur.scriptor.ScriptorMod
import net.minecraft.core.particles.ParticleType
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.StreamCodec

object ScriptorParticles {
  val MAGIC = ScriptorMod.registerParticleType("magic") {
    object: ParticleType<MagicParticleData>(true) {
      override fun codec(): MapCodec<MagicParticleData> {
        return MagicParticleData.CODEC
      }

      override fun streamCodec(): StreamCodec<in RegistryFriendlyByteBuf, MagicParticleData> {
        return MagicParticleData.STREAM_CODEC
      }
    }
  }

  fun register() {}
}
